### PR TITLE
Add image captioning example using MSCOCO

### PR DESCRIPTION
**Pull request by hvy**
Adds an image captioning example with MSCOCO downloader/wrapper.

This implementation includes two variations of the model, one using traditional `LSTM` layers and another one using `NStepLSTM` layers.

Please feel free to comment.

~- [ ] Add beam search.~ (See in #4089 )
- [x] Fix docs and clean up the code.
- [x] Add README.

### CODE REVIEW

**Comment by unnonouno** [comment]
I think beam search is more general technique for NLP. It must be imported as a part of chainer instead of an example code.

**Comment by hvy** [comment]
Do you mean we should exclude beam search from this PR and discuss it elsewhere? I agree that beam search is a more general tool to be reused by others.

**Comment by hvy** [comment]
Reference to previous similar discussion on beam search. https://github.com/chainer/chainer/pull/2344

**Comment by hvy** [comment]
Ready for review.

File: examples/image_captioning/model.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        caption_length = captions.shape[1]
+        for i in range(caption_length - 1):
+            # Compute the loss based on the prediction of the next token in the
+            # sequence
+            x = Variable(self.xp.asarray(captions[:, i]))
+            t = Variable(self.xp.asarray(captions[:, i + 1]))
+            if (t.data == self.embed_word.ignore_label).all():
+                # The next tokens in the sequence should not be accounted for
+                # in the loss, e.g. the targets are paddings that were used to
+                # align the lengths of all captions within the same batch
```
Comment (soskek) [inline examples/image_captioning/model.py]: Could you clarify these comment?
e.g.
Preprocessed captions are padded to reach a maximum length. So, if all targets are pads, prediction is no longer required.

File: examples/image_captioning/model.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        return captions
+
+    def reset(self, img_feats):
+        """Batch of image features to hidden representations.
+
+        Also, reset and then update the internal state of the LSTM.
+        """
+        self.lstm.reset_state()
+        h = self.embed_img(img_feats)
+        h = self.lstm(F.dropout(h, ratio=self.dropout_ratio))
```
Comment (soskek) [inline examples/image_captioning/model.py]: This dropout is redundant.
img_feats already has been through dropout in `VGG16FeatureExtractor`.
In my opinion, it looks good to remove dropout in `VGG16FeatureExtractor` with remaining this line.

File: examples/image_captioning/model.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            self.embed_word = L.EmbedID(
+                vocab_size,
+                hidden_size,
+                initialW=initializers.Normal(1.0),
+                ignore_label=ignore_label
+            )
+            self.embed_img = L.Linear(
+                hidden_size,
+                initialW=initializers.Normal(0.01),
+                initial_bias=initializers.Zero()
```
Comment (soskek) [inline examples/image_captioning/model.py]: As well as I mentioned above, this line is not required.

File: examples/image_captioning/model.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            )
+            self.lstm = L.NStepLSTM(1, hidden_size, hidden_size, dropout_ratio)
+            self.decode_caption = L.Linear(
+                hidden_size,
+                vocab_size,
+                initialW=initializers.Normal(0.01),
+                initial_bias=initializers.Zero()
```
Comment (soskek) [inline examples/image_captioning/model.py]: As well as above.

File: examples/image_captioning/model.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        """Batch of image features and image captions to a singe loss.
+
+        Compute the softmax cross-entropy captioning loss in a single pass
+        without iterating over the sequences.
+        """
+        hx, cx, _ = self.reset(img_feats)
+
+        # Extract all inputs and targets for all captions in the batch
+        xs = [c[:-1] for c in captions]
+        ts = [c[1:] for c in captions]
```
Comment (soskek) [inline examples/image_captioning/model.py]: How about add comments?
```
        xs = [c[:-1] for c in captions]  # del eos
        ts = [c[1:] for c in captions]  # del bos
```

File: examples/image_captioning/model.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+        # Concatenate all input captions and pass them through the model in a
+        # single pass
+        caption_lens = [len(x) for x in xs]
+        caption_sections = np.cumsum(caption_lens[:-1])
+        xs = F.concat(xs, axis=0)
```
Comment (soskek) [inline examples/image_captioning/model.py]: In my opinion, it seems better that `self.step` includes these 3 lines of concat processing of word indices.
Taking linearly concatenated xs is little unnatural as a function.

File: examples/image_captioning/model.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        return loss
+
+    def predict(self, img_feats, bos, eos, max_caption_length):
+        """Batch of image features to captions."""
+        hx, cx, _ = self.reset(img_feats)
+
+        captions = self.xp.full((img_feats.shape[0], 1), bos, dtype=np.int32)
+        for i in range(max_caption_length):
+            xs = Variable(captions[:, -1])
+            hx, cx, ys = self.step(hx, cx, xs, sections=xs.shape[0])
```
Comment (soskek) [inline examples/image_captioning/model.py]: (If you change what I mentioned above, pls follow the lines too.)

File: examples/image_captioning/train.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    parser.add_argument('--mscoco-root', type=str, default='data')
+    parser.add_argument('--max-iters', type=int, default=50000)
+    parser.add_argument('--batch-size', type=int, default=128)
+    parser.add_argument('--dropout-ratio', type=float, default=0.5)
+    parser.add_argument('--val-n-imgs', type=int, default=1000)
+    parser.add_argument('--val-iter', type=int, default=100)
+    parser.add_argument('--log-iter', type=int, default=1)
+    parser.add_argument('--snapshot-iter', type=int, default=1000)
+    parser.add_argument('--rnn', type=str, default='lstm',
+                        choices=['lstm', 'nsteplstm'])
```
Comment (soskek) [inline examples/image_captioning/train.py]: How about using `nsteplstm` as default?

File: examples/image_captioning/train.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    args = parser.parse_args()
+
+    # Load the MSCOCO dataset. Assumes that the dataset has been downloaded
+    # already using e.g. the `download.py` script
+    train, val = datasets.get_mscoco(args.mscoco_root)
+
+    # Use any number of samples from the validation set for validation, using
+    # all of them is usually quite slow since the number of validation images
+    # are many
+    val = val[:args.val_n_imgs]
```
Comment (soskek) [inline examples/image_captioning/train.py]: These lines are confusing now. This default `args.val_n_imgs` equals to `len(val)` (and this line is meaningless now)?
How about using a constant value you specify?
Or, slicing like `val = val[::10]` also looks good. It represents the quantity of dataset to be used intuitively, in this case, 1/10.

File: examples/image_captioning/train.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+class TestModeEvaluator(extensions.Evaluator):
+
+    """Evaluates the model on the validation dataset."""
+
+    def evaluate(self):
+        with chainer.using_config('train', False), \
+                chainer.no_backprop_mode():
+            ret = super(TestModeEvaluator, self).evaluate()
+        return ret
```
Comment (soskek) [inline examples/image_captioning/train.py]: This `TestModeEvaluator` may not required, in the view of 
`chainer.using_config('train', False)` and `chainer.no_backprop_mode()`.
These are performed even in the common `extensions.Evaluator`.
https://github.com/chainer/chainer/blob/v3.2.0/chainer/training/extensions/evaluator.py#L136-L137
https://github.com/chainer/chainer/blob/v3.2.0/chainer/training/extensions/evaluator.py#L182

Or, you have other idea about this class?

File: examples/image_captioning/datasets.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    word_counts = defaultdict(int)
+    for c in captions:
+        for w in split(c):
+            word_counts[w] += 1
+
+    # This vocabulary is needed in order to convert the words in the captions
+    # to integer tokens. When generating captions during testing, these tokens
+    # are mapped back to their corresponding words.
+    vocab = {'<bos>': _bos, '<eos>': _eos, '<unk>': _unk}
+    for w, count in six.iteritems(word_counts):
```
Comment (soskek) [inline examples/image_captioning/datasets.py]: Please check if this works deterministically. If not, word-to-idx dictionary `vocab` can be changed even using same dataset.
The current implementation of `predict.py` uses a vocabulary which is re-constructed by reading dataset. So, if not deterministic, vocabulary used in the `predict.py` can be different from a vocabulary used during training a loaded model.

For example, I think the processing below is deterministic.

```
for w, count in sorted(word_counts.items(), lambda x:(-x[1], x[0])):
```

File: examples/image_captioning/README.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
+# Image Captioning with Convolutional Neural Networks
+
+This is an example implementation of Show and Tell: A Neural Image Caption Generator (https://arxiv.org/abs/1411.4555) a generative image captioning model using a neural network with convolutional and recurrent layers. Given an image, this model generates a sentence that describe it.
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: Fixed.

File: examples/image_captioning/README.md [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+
+## Requirements
+
+This example requires
+
+- [pycocotools](https://github.com/cocodataset/cocoapi/tree/master/PythonAPI)
+
+to work with the dataset. To install pycocotools, clone the repository and run `pip install -e .` from the directory where `setup.py` is located.
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: Rephrased the sentence accordingly.

File: examples/image_captioning/README.md [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+## Model Overview
+
+The model takes an image as input which is fed through a pretrained VGG16 model in order to extract features. These features are then passed to a language model, a recurrent neural network that generates a caption word-by-word until the `EOS` (end-of-sentence) token is encountered or the caption reaches a predetermined maximum caption length. The internals of the language models is a neural network with [LSTM](http://docs.chainer.org/en/stable/reference/generated/chainer.links.LSTM.html) layers. However, Chainer also has a [NStepLSTM](http://docs.chainer.org/en/stable/reference/generated/chainer.links.NStepLSTM.html) layer which does not require sequential passes (for-loops in the code) which is faster. Using the latter, you do not have to align the caption lengths in the training data neither, which you usually do if using the former. This example includes both LSTM and NStepLSTM implementations and preprocessing of the captions. During training, the loss is the softmax cross entropy of correctly predicting the next word in the caption given the current word averaged over the minibatch.
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: Fixed accordingly. Regarding the paragraph being too long. I split it in half and added a header to the second part `More About the Language Model`. Do you think this looks okay?

File: examples/image_captioning/README.md [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+## Dataset
+
+You need to download the MSCOCO captioning dataset in order to train the model. Run the following command to download and extract the necessary files.
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: Fixed.

File: examples/image_captioning/README.md [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+## Training
+
+Once `download.py` finishes, you can start training the model.
+
+```bash
+$ python train.py --rnn nsteplstm --max-caption-length 30 --snapshot-iter 1000 --max-iters 50000 --batch-size 128 --gpu 0
+```
+
+The above example start the training with the NStepLSTM layers in the language model and saves a snapshot of the trained model every 1000 iteration. By default, the first model snapshot is saved as `result/model_1000`.
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: Fixed.

**Comment by hvy** [comment]
Thank you for the comments. I fixed the README for now. 

Regarding explicitly setting 0 biases and wrapping ndarrays in Variables, I think we can skip the former since weights are initialized (and thus demonstrates how you can do this) but keep the latter since the automatic wrapping of ndarrays isn't very intuitive at first.

**Comment by soskek** [comment]
>since the automatic wrapping of ndarrays isn't very intuitive at first.

I got it. OK!

File: examples/image_captioning/datasets.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+import numpy as np
+from PIL import Image
+from pycocotools.coco import COCO
+
+from chainer import dataset
+from chainer.dataset.convert import to_device
+
+
+# Vocabulary tokens of BOS (beginning of sentence), EOS (end of sentence),
+# UNK (unknown word) and token labels to be ignore in the loss computation by
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/datasets.py]: Thanks, fixed.

File: examples/image_captioning/datasets.py [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+    A vocabulary is dynamically created based on all captions and is
+    returned as members of the training and validation dataset objects.
+    """
+    train = MsCocoDataset(root_dir, train_dir, train_anno)
+    val = MsCocoDataset(root_dir, val_dir, val_anno)
+
+    # Create a vocabulary based on the captions from both the training and the
+    # validation datasets
+    anns = train.anns + val.anns
+    captions = [ann['caption'] for ann in anns]
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/datasets.py]: Oh, I didn't know that, thanks. Fixed accordingly.

File: examples/image_captioning/datasets.py (hunk 11 above)
Comment (hvy) [reply examples/image_captioning/datasets.py]: I think `sorted(word_counts.items())` is sufficient to sort the vocab alphanumerically. Updated the code to do so.

File: examples/image_captioning/model.py [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+
+        # Report the loss so that it can be printed, logged and plotted by
+        # other trainer extensions
+        reporter.report({'loss': loss}, self)
+
+        return loss
+
+    def predict(self, imgs, bos, eos, max_caption_length):
+        """Batch of images to captions."""
+        imgs = Variable(imgs)
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/model.py]: Keeping as agreed.

File: examples/image_captioning/model.py [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+            self.embed_word = L.EmbedID(
+                vocab_size,
+                hidden_size,
+                initialW=initializers.Normal(1.0),
+                ignore_label=ignore_label
+            )
+            self.embed_img = L.Linear(
+                hidden_size,
+                initialW=initializers.Normal(0.01),
+                initial_bias=initializers.Zero()
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/model.py]: Removed as agreed.

File: examples/image_captioning/model.py [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+            )
+            self.lstm = L.LSTM(hidden_size, hidden_size)
+            self.out_word = L.Linear(
+                hidden_size,
+                vocab_size,
+                initialW=initializers.Normal(0.01),
+                initial_bias=initializers.Zero()
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/model.py]: Removed as agreed.

File: examples/image_captioning/model.py [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+        self.reset(img_feats)
+
+        loss = 0
+        size = 0
+        caption_length = captions.shape[1]
+        for i in range(caption_length - 1):
+            # Compute the loss based on the prediction of the next token in the
+            # sequence
+            x = Variable(self.xp.asarray(captions[:, i]))
+            t = Variable(self.xp.asarray(captions[:, i + 1]))
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/model.py]: Keeping as agreed.

File: examples/image_captioning/model.py [hunk 23]
```diff
@@ -0,0 +1,267 @@
...
+            if (t.data == self.embed_word.ignore_label).all():
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/model.py]: Ditto.

File: examples/image_captioning/model.py (hunk 3 above)
Comment (hvy) [reply examples/image_captioning/model.py]: Removed as agreed.

File: examples/image_captioning/model.py (hunk 4 above)
Comment (hvy) [reply examples/image_captioning/model.py]: Removed as agreed.

File: examples/image_captioning/train.py (hunk 8 above)
Comment (hvy) [reply examples/image_captioning/train.py]: Fixed for both the training and predicting scripts.

File: examples/image_captioning/model.py [hunk 24]
```diff
@@ -0,0 +1,267 @@
...
+        self.cnn_layer_name = 'fc7'
+
+    def prepare(self, img):
+        """Single image to resized and normalized image."""
+        return L.model.vision.vgg.prepare(img)
+
+    def __call__(self, imgs):
+        """Batch of images to image features."""
+        h = self.cnn(imgs, [self.cnn_layer_name])[self.cnn_layer_name]
+        img_feat = F.dropout(F.relu(h), ratio=0.5)
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/model.py]: Ah, thank you. Then I suppose dropout should be removed as well.

File: examples/image_captioning/model.py [hunk 25]
```diff
@@ -0,0 +1,267 @@
...
+        return self.feat_extractor.prepare(img)
+
+    def __call__(self, imgs, captions):
+        """Batch of images to a single loss."""
+        imgs = Variable(imgs)
+        if self.finetune_feat_extractor:
+            img_feats = self.feat_extractor(imgs)
+        else:
+            # Extract features without building a computational chain
+            with chainer.no_backprop_mode():
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/model.py]: Thanks, will fix.

File: examples/image_captioning/model.py (hunk 1 above)
Comment (hvy) [reply examples/image_captioning/model.py]: Will fix.

File: examples/image_captioning/model.py (hunk 2 above)
Comment (hvy) [reply examples/image_captioning/model.py]: Thank you for the input. I'll keep this line unchanged and remove the ReLU and dropouts from the `VGG16FeatureExtractor `.

File: examples/image_captioning/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (soskek) [reply examples/image_captioning/README.md]: Looks great. I agree with the separation.

File: examples/image_captioning/datasets.py (hunk 11 above)
Comment (soskek) [reply examples/image_captioning/datasets.py]: OK!

File: examples/image_captioning/model.py (hunk 5 above)
Comment (hvy) [reply examples/image_captioning/model.py]: Will fix.

File: examples/image_captioning/model.py (hunk 6 above)
Comment (hvy) [reply examples/image_captioning/model.py]: I'm not too confident about this but is this close to what you meant? https://github.com/chainer/chainer/pull/4076/commits/bdc88f5fc78106a671c6d900a9dce47260178cf1

Comment (soskek) [reply examples/image_captioning/model.py]: yes!! exactly same as i think!

File: examples/image_captioning/train.py (hunk 9 above)
Comment (hvy) [reply examples/image_captioning/train.py]: Changed the argument to `args.val_keep_quantity` and implemented the latter you mentioned. (Also updated comment)

File: examples/image_captioning/train.py (hunk 10 above)
Comment (hvy) [reply examples/image_captioning/train.py]: Thank you for pointing that out! Removed the example-defined class!

**Comment by hvy** [comment]
@soskek I think I covered your feedbacks except the pre-trained model, which I am working on right now (training). Does it look okay to you?

File: examples/image_captioning/README.md [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+
+If you have specified a download directory for MSCOCO when preparing the dataset, add the `--mscoco-root` option followed by the path to that directory.
+
+## Testing
+
+To generate captions for new images, you need to have a snapshot of a trained model. Assuming we are using the model snapshots from the training example after 20000 iterations, we can generate new captions as follows.
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: Training at the moment!

**Comment by soskek** [comment]
The comments are satisfied. Thank you!

How about using a fast iterator like `MultiprocessIterator` instead of serial iterator? If it works well, this example become better also in the view of `MultiprocessIterator` usage.

One more thing is about a place of dataset. I am not sure the downloaded dataset should be stored in a current directory, `$HOME/.chainer/dataset/pfnet/chainer` or others, which may depend on policy of Chainer development.

File: examples/image_captioning/datasets.py [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+        self.coco_data = data_dir
+
+    def __len__(self):
+        return len(self.anns)
+
+    def get_example(self, i):
+        """Called by the iterator to fetch a data sample.
+
+        A data sample from MSCOCO consists of an image and its corresponding
+        caption.
```
Comment (kmaehashi) [inline examples/image_captioning/datasets.py]: I think it is nice if the shape of returned image is described here.

File: examples/image_captioning/model.py [hunk 28]
```diff
@@ -0,0 +1,283 @@
...
+    """VGG16 image feature extractor."""
+
+    def __init__(self):
+        super(VGG16FeatureExtractor, self).__init__()
+        with self.init_scope():
+            self.cnn = L.VGG16Layers()
+        self.cnn_layer_name = 'fc7'
+
+    def prepare(self, img):
+        """Single image to resized and normalized image."""
```
Comment (kmaehashi) [inline examples/image_captioning/model.py]: Could you explain that RGB-to-BGR conversion is performed here?
It is hard for beginners to guess what `L.model.vision.vgg.prepare` does at a glance.

File: examples/image_captioning/train.py [hunk 29]
```diff
@@ -0,0 +1,131 @@
+import argparse
```
Comment (kmaehashi) [inline examples/image_captioning/train.py]: Please add execution bit (`chmod +x`) and shebang line for scripts expected to be run directly.
(Also for `predict.py` and `download.py`)

File: examples/image_captioning/README.md [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+This downloads and extracts the training and validation images, as well as necessary meta data including captions to a `data` directory under the current folder.
+You can change the output directory by appeding `--out` followed by the target directory.
+Notice that this may take a while and that it requires approximately 20 GB of disk space.
+
+## Training
+
+Once `download.py` finishes, you can start training the model.
+
+```bash
+$ python train.py --rnn nsteplstm --max-caption-length 30 --snapshot-iter 1000 --max-iters 50000 --batch-size 128 --gpu 0
```
[the remark this answers is not in the document]
Comment (kmaehashi) [reply examples/image_captioning/README.md]: One more thing; could you please mention `export MPLBACKEND=Agg` like as in `vae` example? I failed to run the training in headless environment.

**Comment by kmaehashi** [comment]
Reminder: after merging this I have to add test code to https://github.com/chainer/chainer-test/blob/master/test_example.sh

**Comment by hvy** [comment]
@soskek , @kmaehashi thanks for the additional comments. I'll go through them Tue, Wed.

File: examples/image_captioning/README.md [hunk 31]
```diff
@@ -0,0 +1,63 @@
+# Image Captioning with Convolutional Neural Networks
+
+This is an example implementation of Show and Tell: A Neural Image Caption Generator (https://arxiv.org/abs/1411.4555) a generative image captioning model using a neural network with convolutional and recurrent layers.
+Given an image, this model generates a sentence that describes it.
+
+
+## Requirements
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: Added!

File: examples/image_captioning/README.md [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+
+The model takes an image as input which is fed through a pretrained VGG16 model in order to extract features.
+These features are then passed to a language model, a recurrent neural network that generates a caption word-by-word until the `EOS` (end-of-sentence) token is encountered or the caption reaches a maximum length.
+During training, the loss is the softmax cross entropy of predicting the next word given the preceding words in the caption.
+
+### More About the Language Model
+
+The internals of the language models is a neural network with [LSTM](http://docs.chainer.org/en/stable/reference/generated/chainer.links.LSTM.html) layers.
+However, Chainer also has a [NStepLSTM](http://docs.chainer.org/en/stable/reference/generated/chainer.links.NStepLSTM.html) layer which does not require sequential passes (for-loops in the code) which is faster. Using the latter, you do not have to align the caption lengths in the training data neither, which you usually do if using the former.
+This example includes both LSTM and NStepLSTM implementations and preprocessing of the captions.
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: I think the difference is explained in terms of performance and how data is preprocessed. Added some more explanations though. What do you think?

File: examples/image_captioning/README.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: You are right that `--max-caption-length` has no effect when training with NStepLSTM. Added an explanation for this and the Agg backend.

File: examples/image_captioning/README.md [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+
+If you have specified a download directory for MSCOCO when preparing the dataset, add the `--mscoco-root` option followed by the path to that directory.
+
+## Testing
+
+To generate captions for new images, you need to have a snapshot of a trained model.
+Assuming we are using the model snapshots from the training example after 20000 iterations, we can generate new captions as follows.
+
+```bash
+$ python predict.py --img cat.jpg --model result/model_20000 --rnn nsteplstm --max-caption-length 30 --gpu 0
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/README.md]: Fixed!

File: examples/image_captioning/datasets.py (hunk 27 above)
Comment (hvy) [reply examples/image_captioning/datasets.py]: Fixed.

File: examples/image_captioning/train.py (hunk 29 above)
Comment (hvy) [reply examples/image_captioning/train.py]: Fixed.

File: examples/image_captioning/train.py [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+from chainer import training
+from chainer.training import extensions
+
+import datasets
+from model import ImageCaptionModel
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--out', type=str, default='result')
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/train.py]: Fixed.

File: examples/image_captioning/train.py [hunk 35]
```diff
@@ -0,0 +1,131 @@
...
+        model.to_gpu()
+
+    def transform(in_data):
+        # Called for each sample and applies necessary preprocessing to the
+        # image such as resizing and normalizing
+        img, caption = in_data
+        img = model.prepare(img)
+        return img, caption
+
+    # Usually not required but we need to preprocess the images since their
```
[the remark this answers is not in the document]
Comment (hvy) [reply examples/image_captioning/train.py]: Fixed. Removed the first part.

**Comment by hvy** [comment]
Change to using `MultiprocessIterator` instead of `SerialIterator`.

**Comment by hvy** [comment]
After offline discussions we concluded that it is okay to use a separate script (whether or not using the `download` utility) for preparing the dataset (as long as the user is able to specify the target directory), when the dataset is larger. @soskek do you think the current implementation looks okay?

**Comment by soskek** [comment]
OK!!

**Comment by kmaehashi** [comment]
LGTM!

**Pull request merged**